csrc/iter_visitor.h

```diff
@@ -372,4 +370,146 @@ class TORCH_CUDA_CU_API InputsOf : public IterVisitor {
+  void handle(Expr* expr) override;
+
+  //! We implement this in order to remove dangling TensorViews whose uses are
+  //! all dead. Note that we do not remove other ValTypes like Scalars since
+  //! they might still be used as attributes or members of other objects, which
+  //! is not reflected by Val::uses().
+  void handle(TensorView* tv) override;
+
+  //! Replaces a Val in outputs, and in all uses.
+  //!
+  //! The argument old_val is always marked Dead by this method. If old_val is a
+  //! Fusion input, we do not replace it. If old_val's definition is non-null
+  //! and has other outputs which are not dead, we do not remove old_val.
+  //!
+  //! Returns whether old_val was registered for removal from the Fusion.
+  bool replaceVal(Val* old_val, Val* new_val);
+
+  //! Find whether a statement is not marked as live code.
+  inline bool isDead(Statement* stmt) const {
+    return live_statements_.find(stmt) == live_statements_.end();
+  }
+
+  //! Find whether a statement is marked as live code.
+  inline bool isLive(Statement* stmt) const {
+    return !isDead(stmt);
+  }
+
+  //! Check whether all outputs of an expression have been marked dead
+  inline bool allOutputsDead(Expr* expr) const {
+    return std::all_of(
+        expr->outputs().begin(), expr->outputs().end(), [&](Val* outp) {
+          return isDead(outp);
+        });
+  }
+
+  //! Check whether all uses have been marked dead
+  inline bool allUsesDead(Val* val) const {
+    return std::all_of(val->uses().begin(), val->uses().end(), [&](Expr* use) {
+      return isDead(use);
+    });
+  }
+
+ private:
+  //! Removes an Expr* from the Fusion, if possible.
+  //!
+  //! The Expr will _only_ be marked dead and removed if all of its outputs are
+  //! already marked dead. In this case all the outputs will also be removed
+  //! from the Fusion.
+  //!
+  //! Returns whether the Expr was marked dead and removed from the Fusion.
+  bool maybeRemoveExpr(Expr* expr);
+
+  //! Mark a single Statement as being alive.
+  inline void markLive(Statement* stmt) {
+    live_statements_.insert(stmt);
+  }
+
+  //! Ensure that a Statement and its upstream Statements are alive. If it is an
+  //! Expr, ensure all its inputs are alive. If it's a Val with a definition,
+  //! recursive to the definition. Newly-created Statements default to being
+  //! dead, so this method is called when adding a Statement to the active path
+  //! of the Fusion inside replaceVal.
+  void markLiveRecursive(Statement* stmt);
+
+  //! Mark a single Statement as being dead. This does not remove stmt from the
+  //! Fusion. It is an error to call this on a Fusion output.
+  //!
+  //! Returns true if the statement was previously live, and false otherwise.
+  bool markDead(Statement* stmt);
+
+  //! Register a Val for later removal.
+  void registerRemoval(Val* val);
+
+  //! Register a Val for later replacement
+  inline void registerReplacement(Val* old_val, Val* new_val) {
```

Comment:
Looks like this is just used from `replaceVal`. Since it's just one line, why not just combine them and name it as `registerReplacement`?